conan/tools/cmake/toolchain.py

```diff
@@ -395,28 +395,56 @@ def context(self):
         {% if find_package_prefer_config %}
         set(CMAKE_FIND_PACKAGE_PREFER_CONFIG {{ find_package_prefer_config }})
         {% endif %}
-        # To support the generators based on find_package()
-        {% if cmake_module_path %}
-        set(CMAKE_MODULE_PATH {{ cmake_module_path }} ${CMAKE_MODULE_PATH})
+        {% if generators_folder or host_build_paths_root or host_build_paths_noroot or build_build_paths %}
+        set(CMAKE_MODULE_PATH {{ generators_folder }} {{ host_build_paths_root }} {{ host_build_paths_noroot }} {{ build_build_paths }} ${CMAKE_MODULE_PATH})
         {% endif %}
-        {% if cmake_prefix_path %}
-        set(CMAKE_PREFIX_PATH {{ cmake_prefix_path }} ${CMAKE_PREFIX_PATH})
+        {% if generators_folder or host_build_paths_noroot %}
+        set(CMAKE_PREFIX_PATH {{ generators_folder }} {{ host_build_paths_noroot }} ${CMAKE_PREFIX_PATH})
         {% endif %}
-        {% if android_prefix_path %}
-        set(CMAKE_FIND_ROOT_PATH {{ android_prefix_path }} ${CMAKE_FIND_ROOT_PATH})
+        {% if cmake_program_path %}
+        set(CMAKE_PROGRAM_PATH {{ cmake_program_path }} ${CMAKE_PROGRAM_PATH})
+        {% endif %}
+        {% if cmake_library_path %}
+        set(CMAKE_LIBRARY_PATH {{ cmake_library_path }} ${CMAKE_LIBRARY_PATH})
+        {% endif %}
+        {% if is_apple and cmake_framework_path %}
+        set(CMAKE_FRAMEWORK_PATH {{ cmake_framework_path }} ${CMAKE_FRAMEWORK_PATH})
+        {% endif %}
+        {% if cmake_include_path %}
+        set(CMAKE_INCLUDE_PATH {{ cmake_include_path }} ${CMAKE_INCLUDE_PATH})
         {% endif %}
 
-        # To support cross building to iOS, watchOS and tvOS where CMake looks for config files
-        # only in the system frameworks unless you declare the XXX_DIR variables
-        {% if cross_ios %}
-            set(CMAKE_FIND_ROOT_PATH_MODE_PACKAGE BOTH)
+        {% if cross_building %}
+        if(NOT DEFINED CMAKE_FIND_ROOT_PATH_MODE_PACKAGE OR CMAKE_FIND_ROOT_PATH_MODE_PACKAGE STREQUAL "ONLY")
+            set(CMAKE_FIND_ROOT_PATH_MODE_PACKAGE "BOTH")
+        endif()
+        if(NOT DEFINED CMAKE_FIND_ROOT_PATH_MODE_PROGRAM OR CMAKE_FIND_ROOT_PATH_MODE_PROGRAM STREQUAL "ONLY")
+            set(CMAKE_FIND_ROOT_PATH_MODE_PROGRAM "BOTH")
+        endif()
+        if(NOT DEFINED CMAKE_FIND_ROOT_PATH_MODE_LIBRARY OR CMAKE_FIND_ROOT_PATH_MODE_LIBRARY STREQUAL "ONLY")
+            set(CMAKE_FIND_ROOT_PATH_MODE_LIBRARY "BOTH")
+        endif()
+        {% if is_apple %}
+        if(NOT DEFINED CMAKE_FIND_ROOT_PATH_MODE_FRAMEWORK OR CMAKE_FIND_ROOT_PATH_MODE_FRAMEWORK STREQUAL "ONLY")
+            set(CMAKE_FIND_ROOT_PATH_MODE_FRAMEWORK "BOTH")
+        endif()
         {% endif %}
-        """)
+        if(NOT DEFINED CMAKE_FIND_ROOT_PATH_MODE_INCLUDE OR CMAKE_FIND_ROOT_PATH_MODE_INCLUDE STREQUAL "ONLY")
+            set(CMAKE_FIND_ROOT_PATH_MODE_INCLUDE "BOTH")
+        endif()
+        {% endif %}
+    """)
+
+    @staticmethod
```

Comment:
Hi @SpaceIm, 
I have been talking with the team and discussing how to introduce this PR and the migration to 2.0.  We all understood the issue and agree we must do something.

We have a question regarding this block of code of the `CMAKE_FIND_ROOT_PATH_MODE_*` stuff. We want to simplify it because it is a bit complex and could be problematic. We thought that maybe we could append `{{ generators_folder }} {{ host_build_paths_noroot }}` to the `CMAKE_FIND_ROOT_PATH` always without checking the value of the `CMAKE_FIND_ROOT_PATH_MODE_*` variables? WDYT?

Reply:
It would break `find_program()` of buid context executables if:
- the same recipe is in build_requires and requires
- `CMAKE_FIND_ROOT_PATH_MODE_PROGRAM` is `ONLY`

Reply:
Too fragile.

It would break `find_program()` of build context executables when:
-  `CMAKE_FIND_ROOT_PATH_MODE_PROGRAM` is `ONLY`.
- `CMAKE_FIND_ROOT_PATH_MODE_PROGRAM` is `BOTH` and the same recipe is in requires and build_requires.

Reply:
Yes, this toolchain doesn't change `CMAKE_FIND_ROOT_PATH_MODE_*` if it's `NEVER` or already `BOTH`.
Because all the logic in this PR is very robust to `NEVER` so we are fine with this value.
`BOTH` is also fine because we don't manipulate `CMAKE_FIND_ROOT_PATH`. Of course if user toolchain injects stuff in `CMAKE_FIND_ROOT_PATH` which could take precedence over conan packages, it's up to him.